avl_template_new.py

```diff
@@ -1072,21 +1072,45 @@ def shuffle(self, array):
 	@returns: the absolute value of the difference between the height of the AVL trees joined
 	"""
 	def concat(self, lst):
-		absDiff = abs(self.root.height - lst.root.height)
+		absDiff = abs(self.getRoot().getHeight() - lst.getRoot().getHeight())
+		shorterTree = self
+		LongerTree = lst
+
+		b = LongerTree.getRoot()
+		while b.getHeight() > shorterTree.getRoot().getHeight():
+			b = b.getLeft()
 		
-		LastNodeInSelf = self.retrieveNode(self.size)
-		firstNodeInLst = lst.retrieveNode(self.size)
+		maxInShort = shorterTree.retrieveNode(shorterTree.size - 1)
+		maxInShort.getParent().right = AVLNode.virtualNode()
+
+		# if be is the root of lst
+		if b.getParent() != None:
+			maxInShort.parent = b.getParent()
+			maxInShort.getParent().left = maxInShort
+
+			maxInShort.right = b
+			b.parent = maxInShort
+
+			maxInShort.left = shorterTree.getRoot()
+			maxInShort.setHeight(max(maxInShort.getRight().getHeight(), maxInShort.getLeft().getHeight())+1)
```

Comment:
you can use `maxInShort.updateHelpers()` and it will also set the size correctly